### ISSUE TITLE
Beam search incorrectly sampled 2*num_beams tokens instead of num_beams tokens.

### ISSUE THREAD

**Issue by transcend-0** [huggingface/transformers#34168]
### System Info

Python 3.10.11
transformers > 4.40 (all versions released so far)
torch 2.0.1
Linux version 4.15.0-55-generic x86_64

### Who can help?

@gante 

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

the source code in transformers.generation.utils.py (line 3299 in v4.45.0)

![image](https://github.com/user-attachments/assets/31902081-d37c-4dbb-8cc9-ffc532e6cf29)

### Expected behavior

Why is `max(2, 1 + n_eos_tokens)` used here instead of `max(1, 1 + n_eos_tokens)` ?

This results in most of the time, beam search first samples **_2*num_beams_** tokens, then takes the **_top num_beams_**, rather than directly sampling **_num_beams_** tokens. This operation is different from the beam search we generally consider.

**Comment by gante** [comment]
TL;DR if all beams pick `eos_token_id` as the next token, generation would stop. However, depending on the parameterization, it may still be possible to continue beam search and get a better score (see `beam_scorer.is_done` and how this attribute is set). We want to guarantee there are at least `num_beams` valid continuations for the next iteration, so we can continue beam search to potentially get a better score.
